pkg/cli/tail_test.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cli
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"cloud.google.com/go/logging/apiv2/loggingpb"
+	"github.com/abcxyz/pkg/testutil"
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+func TestTailCommand(t *testing.T) {
+	t.Parallel()
+
+	ct := time.Now().UTC()
+
+	validLog := &loggingpb.LogEntry{
+		InsertId: "test-log",
+		Payload: &loggingpb.LogEntry_JsonPayload{
+			JsonPayload: &structpb.Struct{
+				Fields: map[string]*structpb.Value{
+					"service_name":  structpb.NewStringValue("foo_service"),
+					"method_name":   structpb.NewStringValue("foo_method"),
+					"resource_name": structpb.NewStringValue("foo_resource"),
+					"authentication_info": structpb.NewStructValue(&structpb.Struct{
+						Fields: map[string]*structpb.Value{
+							"principal_email": structpb.NewStringValue("foo@bet.com"),
+						},
+					}),
+				},
+			},
+		},
+		Labels: map[string]string{"environment": "dev", "accessing_process_name": "foo_apn"},
+	}
+
+	validLogJson, err := protojson.Marshal(validLog)
```

Comment:
Two issues with protojson.Marshal:
1. We use protojson.Marshal to marshal the log entry tailed to JSON in tail.go as well, it seems protojson.Marshal does not produce deterministic results.  
2. I always got `failed to marshal log to json (InsertId: "-oojsuhe1sciu"): proto: google.protobuf.Any: unable to resolve "type.googleapis.com/google.iam.v1.logging.AuditData": not found` error when I try to parse a log entry with proto payload. Wondering why and if it is a risk (though we only expect json payload).

Reply:
update:
1. it is known issue - https://github.com/golang/protobuf/issues/1121, going to stripe the spaces in unit test as a workaround.

Reply:
update:
2. We should be good as we only expecting json payload.

Reply:
Yes, we don't expect proto payload